### ISSUE TITLE
4 spaces

### ISSUE THREAD

**Issue by neongreen** [tweag/ormolu#442]
I'll bite this bullet.

There are no _particularly_ good arguments for either 2 spaces or 4 spaces. However, I can think of a few weak ones for 4 spaces:

1. "From a survey of the top downloaded 1000 packages on Hackage, 660 are 4-spaced and 343 are 2-spaced. All else being equal, 4 spaces wins" ([link](https://chrisdone.com/posts/hindent-5/#fn2))

2. 4 spaces discourages extreme nesting, which is probably a good thing.

3. More indentation helps people with not-so-great vision (e.g. me).

4. The author of elm-format (probably the closest to Ormolu formatter out there) [reports some interesting data](https://github.com/avh4/elm-format/issues/158#issuecomment-219262937) (emphasis mine):

   > elm-format initially used 4-spaces, since that was the preferred style at the time. Some people had requested 2-space indentation, so I ran a survey to help me make an informed choice about whether or not to try 2-spaces.
   >
   > Here is what the survey found:
   >
   > * 85% of people have only tried one style
   > * **98% of people who have only tried one style prefer that style**
   > * of the people who have tried both styles, 15% had no preference, and the rest were evenly split between 4-spaces and 2-spaces
   >
   > Since most people had been using 4-space indentation, we decided to try 2-space indentation as an experiment to give it a fair chance. **The feedback I've received from people who have now used both for extended periods of time has been in favor of 4-space indentation due to improved legibility.**

Given all this, I'm proposing to switch to 4-space indentation. (I'm also willing to submit a PR.)

**Comment by int-index** [comment]
I've found it extremely painful to try and fit code into 80 characters using 4 space indentation. The expressions simply don't fit.

And 80 characters are an important limit. It means that two panes of code take 160 characters, and that's about how much characters fit on most laptop screens.

> 4 spaces discourages extreme nesting, which is probably a good thing.

It encourages short, non-descriptive names, so that at higher indentation levels expressions still fit 80 columns.

> More indentation helps people with not-so-great vision (e.g. me).

Bigger fonts help people with not-so-great vision (e.g. me), and to allow bigger fonts, it's better to sacrifice useless indentation.

**Comment by cblp** [comment]
I personally like 4 spaces, but Haskell is very hard to use with 4 spaces in corner cases. Consider:

```hs
( ( a,
    b
  ),
  c
)
```

Do you have any idea of how this could be implemented with 4 spaces?

```hs
(   (   a, -- like this?
        b
    ),
    c
)
```

**Comment by cblp** [comment]
@int-index What if your expressions are too complex? There are programming style guides where 8 spaces are used for indentation. How do you think they manage to fit their expressions in a line?

**Comment by int-index** [comment]
> What if your expressions are too complex?

I don't think they are. Here's one example:

```
renderMotion ::
  LayoutCtx ->
  MotionAction ->
  Paths ->
  FindZone ->
  CairoRender DrawCtx
renderMotion lctx action Paths {pathsSelection} (Find findZone) =
  case findZone (selectionPath pathsSelection) of
    Nothing -> error "renderMotion: no selection"
    Just (o, e) ->
      getNoAnn
        $ foldCairoCollage o
        $ mapCollageAnnotation (const mempty)
        $ motionPreview e
  where
    motionPreview :: Extents -> Collage Ann El
    motionPreview e =
      case action of
        MotionNoOp ->
          substrateMargin (outline 2 motionBorderColor . extentsMax e) $
            rect nothing nothing e
        MotionToggleEditMode ->
          substrateMargin (outline 2 motionBorderColor . extentsMax e) $
            rect nothing nothing e
        MotionPopSwap n ->
          substrate 0 (rect nothing dark1')
            $ substrateMargin (outline 2 motionBorderColor . extentsMax e)
            $ collageWithMargin (Margin 4 4 4 4)
            $ snd (layoutNode lctx' n precAllowAll)
    lctx' = lctx {_lctxValidationResult = mempty}
```

That's fairly straightforward code. And here's what happens when 2-space indentation is replaced with 4-space indentation:

```
renderMotion ::
    LayoutCtx ->
    MotionAction ->
    Paths ->
    FindZone ->
    CairoRender DrawCtx
renderMotion lctx action Paths {pathsSelection} (Find findZone) =
    case findZone (selectionPath pathsSelection) of
        Nothing -> error "renderMotion: no selection"
        Just (o, e) ->
            getNoAnn
                $ foldCairoCollage o
                $ mapCollageAnnotation (const mempty)
                $ motionPreview e
    where
        motionPreview :: Extents -> Collage Ann El
        motionPreview e =
            case action of
                MotionNoOp ->
                    substrateMargin (outline 2 motionBorderColor . extentsMax e) $
                        rect nothing nothing e
                MotionToggleEditMode ->
                    substrateMargin (outline 2 motionBorderColor . extentsMax e) $
                        rect nothing nothing e
                MotionPopSwap n ->
                    substrate 0 (rect nothing dark1')
                        $ substrateMargin (outline 2 motionBorderColor . extentsMax e)
                        $ collageWithMargin (Margin 4 4 4 4)
                        $ snd (layoutNode lctx' n precAllowAll)
        lctx' = lctx {_lctxValidationResult = mempty}
```

It looks fine, but it no longer fits 80 characters. There are two options to make it fit:

* use shorter names (making it less readable)
* factor out some code (making it more complex)

> There are programming style guides where 8 spaces are used for indentation. How do you think they manage to fit their expressions in a line?

No idea.

**Comment by cblp** [comment]
> factor out some code (making it more complex)

I'm pretty sure factoring code makes it simpler generally. The whole construction becomes harder a bit, maybe, but each part becomes way easier to understand if you abstract properly.

So splitting code with abstraction is good in general.

**Comment by int-index** [comment]
> I'm pretty sure factoring code makes it simpler generally.

And I believe inlining it makes it more straightforward. So, may I continue to decide whether to inline or factor out based on the actual code instead of stylistic restrictions? It's much easier with 2 spaces.

**Comment by luke-clifton** [comment]
> There are two options to make it fit:
> * use shorter names (making it less readable)
> * factor out some code (making it more complex)

<details><summary>
Or... just reformat the code slightly?
</summary>

```haskell
renderMotion ::
    LayoutCtx ->
    MotionAction ->
    Paths ->
    FindZone ->
    CairoRender DrawCtx
renderMotion lctx action Paths {pathsSelection} (Find findZone) =
    case findZone (selectionPath pathsSelection) of
        Nothing -> error "renderMotion: no selection"
        Just (o, e) ->
            getNoAnn
                $ foldCairoCollage o
                $ mapCollageAnnotation (const mempty)
                $ motionPreview e
    where
        motionPreview :: Extents -> Collage Ann El
        motionPreview e =
            case action of
                MotionNoOp ->
                    substrateMargin
			(outline 2 motionBorderColor . extentsMax e)
                        (rect nothing nothing e)
                MotionToggleEditMode ->
                    substrateMargin
			(outline 2 motionBorderColor . extentsMax e)
                        (rect nothing nothing e)
                MotionPopSwap n ->
                    substrate 0 (rect nothing dark1')
                        $ substrateMargin 
			    (outline 2 motionBorderColor . extentsMax e)
                        $ collageWithMargin (Margin 4 4 4 4)
                        $ snd (layoutNode lctx' n precAllowAll)
        lctx' = lctx {_lctxValidationResult = mempty}
```
</details>

Personally, if I have to spend time in a 2-space code base I usually end up substituting all occurrences of two spaces with four spaces and back again if I have to make changes. I have no trouble keeping within 80 columns with four spaces. I think 2 spaces is a real accessibility issue, and my eyesight isn't even that bad yet!

**Comment by pwm** [comment]
This is an inherently subjective topic.

I personally used to use 4 spaces for a long time in the past and then switched to 2 spaces years ago which i've been using and preferring ever since. To me the win with 2 spaces are freed up vertical real-estate, same argument as @int-index , and also increased information density which I find preferable. All this is totally subjective though.

As a side-note, and I'm sure I'm stating the obvious here, tabs were invented to deal with exactly this problem but Haskell being layout sensitive tabs are a no go so we're stuck with having to make choices like this one.

**Comment by mrkkrp** [comment]
It makes sense to me to keep this open just to let people express their opinions. So, thanks @neongreen for opening it.

2 spaces indentation is objectively better for reasons @cblp and @int-index already mentioned. Especially the ability to fit more on a line without exceeding 80 character limit and the ablility to achieve deeper levels of nesting is very important.

I find the counter argument pertinent to the "poor vision" issues, which was brought up several times, first time by @chshersh on Reddit, not convincing. I do have poor vision and I do wear glasses. If I find code hard to read, I increase the font size. It is as simple as that. Haskell is prone to having complex expressions which will force you to parse/pay attention to little details anyway. It is not like we only have some simple structure where indentation levels define all the meaning. In that case maybe increasing indentation step would solve the issue and you could avoid increasing font size.

Also, there are a few languages like Lisps where you traditionally have 2 spaces indentation and no one argues about that. People with poor vision write code in those languages, too. Health issues is a sympathetic reason, but every time one tries to leverage that to change our decisions I feel like being manipulated.

**Comment by neongreen** [comment]
"poor vision" is just one of the arguments, and I would say not the most
interesting one.

**Comment by neongreen** [comment]
(plus, Vlad's argument regarding font size is just as valid, yeah)

**Comment by chshersh** [comment]
It's not like my opinion matters a lot here. But because I was asked to provide motivation in favour of 4 spaces, I will write my thoughts. Of course, this is a subjective topic. But I prefer 4 spaces and leading punctuation. This style makes code more comfortable for me to read. I got used to it, and I don't think I will ever switch to 2 spaces (unless I'm getting paid for that). Everything was already sad, so I will mostly repeat that:

* 4 spaces discourage deep nestedness. If you have lines with 20 spaces indentation (level 5), it is an excellent sign to refactor something. Usually, you have deep levels of nestedness with nested `case`s and `where` blocks. Those definitions reuse a lot of logic from the context. I like Haskell because it allows me to keep as little required context as possible in my head when reading a definition of a function. So it might be a good idea to move something out to reduce the scope of the context. And 4 spaces encourage to move small independent blocks more often.
* I try to give descriptive variable names when possible. Giving proper variable names in programming is a challenge for everyone, but I usually don't introduce short variable names just to fit the line length limit. Because when it becomes relevant (on deep levels), it's not possible to have a short and descriptive name. So, again, I'm just moving the code out.

Regarding example by @int-index, I would personally format it like this:

```haskell
renderMotion
    :: LayoutCtx
    -> MotionAction
    -> Paths
    -> FindZone
    -> CairoRender DrawCtx
renderMotion lctx action Paths{pathsSelection} (Find findZone) =
    case findZone (selectionPath pathsSelection) of
        Nothing -> error "renderMotion: no selection"
        Just (o, e) -> getNoAnn
            $ foldCairoCollage o
            $ mapCollageAnnotation (const mempty)
            $ motionPreview e
  where
    motionPreview :: Extents -> Collage Ann El
    motionPreview e = case action of
        MotionNoOp -> substrateMargin
            (outline 2 motionBorderColor . extentsMax e)
            (rect nothing nothing e)
        MotionToggleEditMode -> substrateMargin
            (outline 2 motionBorderColor . extentsMax e)
            (rect nothing nothing e)
        MotionPopSwap n -> substrate 0 (rect nothing dark1')
            $ substrateMargin (outline 2 motionBorderColor . extentsMax e)
            $ collageWithMargin (Margin 4 4 4 4)
            $ snd (layoutNode lctx' n precAllowAll)

    lctx' :: LayoutCtx
    lctx' = lctx {_lctxValidationResult = mempty}
```

Each line still fits into 80 character length limit. But to my taste, it feels better. You can also notice that I'm not a fan of vertical code alignment. I don't see any need to move `case`s on separate lines because they don't help to minimise diffs. And I don't have problems with git diffs. The style guide I'm following suggests to minimise `git` diffs but only in the aspect when nestedness mustn't depend on the length of preceding identifiers.

**Comment by mrkkrp** [comment]
@chshersh Thanks for the input. It's not a totally unreasonable position, but all things considered, the user should be able to have deeply nested constructions if he/she decides that it is necessary without being forced to introduce auxiliary definitions. I'm willing to trust the user, so that if he/she wants to nest things, I'd say he/she has reasons to do it.

> t's not like my opinion matters a lot here.

If you define "to matter" as "to be accepted as the only truth without a debate". Otherwise this statement is not fair. It often happens in life that your point of view may differ from what other people think and you are unable to persuade them. It doesn't mean that the system is automatically unfair or you're being ignored.

**Comment by chshersh** [comment]
> I find the counter argument pertinent to the "poor vision" issues, which was brought up several times, first time by @chshersh on Reddit, not convincing.

> ... every time one tries to leverage that to change our decisions I feel like being manipulated.

As to non-technical aspects of this issue, it wasn't me who brought this idea to `ormolu`. I never told `ormolu` maintainers how they should implement their tool, I was only describing my reasons why I wouldn't use this particular formatter in the non-related to the project thread without asking for any actions. Bringing me to this thread, pointing to my health issues in the same sentence and choosing me specifically among all people (who **actually** decided to express their opinion here) is very inappropriate behaviour. I haven't opened this issue, and I haven't used my poor vision as a reason to change how this tool behaves. But yet I'm somehow the one to be the ultimate evil incarnate trying to manipulate you (what?). You just decided to point to me specifically and tell publically that "my arguments are not convincing" (using only minor one of them though) when I haven't even participated in this discussion!

Please, make sure you are not making your tool maintenance discussion personal. I will just quote your words here:

> It is as simple as that.

I don't think that it makes sense for me to participate in this discussion any further. However, I wish you luck with the project anyway.

**Comment by mrkkrp** [comment]
I'm sincerely sorry if my replies indeed read like this, this wasn't the intent. The reason I used the word *manipulate* is because it is easy to use health issues (a topic many people including me will sympathize) as a strong argument where in this particular case it's not really a good enough reason to change indentation step. Another reason I used the word manipulate is because you tried to make portray yourself as a victim (which is why I remember you now among all people), iirc [you used the phrase](https://www.reddit.com/r/haskell/comments/dgdber/tweag_io_ormolu_announcing_first_release/f3czh3y?utm_source=share&utm_medium=web2x):

>  So fuck me because I'm not good enough to use this tool

I agree that the phrase "it is as simple as that" may be read as non-neutral (even though all I meant is to underline that the solution is simple), and for that I apologize. But what you said in the original Reddit thread is not a neutral behavior either, as well as casually droping phrases like "It's not like my opinion matters a lot here", which is also a form of manipulation ("he is feeling unwelcome, so let's just do what he wants to rectify"). I do not like when people use this sort of strategy in a discussion. I'd like us to stick to the facts, or opinion, but neutrally expressed.

**Comment by luke-clifton** [comment]
I'll agree that it is perhaps easier to write code with 2-space indentation. I find it harder to read though. And I tend to err on the side of readability rather than ease of writing.

> every time one tries to leverage that to change our decisions I feel like being manipulated.

It's an argument like any other. The fact that you feel manipulated seems to me like you don't think it's true, and that I am trying to guilt you into taking my position. I'm not. I genuinely find it difficult to follow the structure as I'm sure many other people do as well.

Changing the font size doesn't really work as a solution, as you just end up moving code off screen which makes it even more difficult to see the structure as it's not even on the screen at the same time.

Perhaps it would be best if ormolu simply took a flag for this. I realise it goes against a "one true format", but as long as we can make it roundtrip between the two, it becomes trivial for people to configure their editors to show the formatting they desire, and write the formatting the project desires.
As long as changing the indentation doesn't change line breaks I don't see why this would be a problem.

I also think it would improve adoption. There are many formatting decisions that ormolu makes that I disagree with, but I can live with them, 2-space indentation is not one of them, and I'm sure I'm not the only one.

----

> As a side-note, and I'm sure I'm stating the obvious here, tabs were invented to deal with exactly this problem but Haskell being layout sensitive tabs are a no go so we're stuck with having to make choices like this one.

Tabs work fine in Haskell. I used to use them exclusively for this very reason (though, people still argue about the 80 character thing, and whether it's 80 characters with 2 space or 4 space tabs). I switched to spaces after GHC started warning by default on tabs because it became to much hassle.

**Comment by mrkkrp** [comment]
@luke-clifton I have poor vision too and I never had any problems with 2 spaced indentation. All of this all is purely subjective and is not connected to vision issues at all.

No one commented though, should we be writing this:

```haskell
(   (   a, -- like this?
        b
    ),
    c
)
```

Because every line should start at multiple of indentation step. 2 spaces work naturally here, while 4 spaces just look terrible.

**Comment by mboes** [comment]
There's a difference between alignment and indentation. You might say, "but we never align", but we do. It's only that we don't align in ways that depend on the length of identifiers in previous lines. In the snippet above, the second line is aligned with the `a` on the previous line, no matter at what column the `a` is.

**Comment by cblp** [comment]
In the 4-space paradigm, we can switch to
```hs
(
    (
        a,
        b
    ),
    c
)
```

but somebody will complain now about the number of lines.

**Comment by mrkkrp** [comment]
@mboes We do align, but right now every line starts at multiple of 2, we arranged it so. It is nice when your editor supports something like vertical bars to show you different indentation steps. It's less nice when you code goes out-of-sync with the bars.

**Comment by wraithm** [comment]
Just throwing in my two cents, 4 spaces wins for me because of the deeply nesting issue. I find it harder to read 2 space indented code, even with bumping up the font size. It's not a matter of difficulty reading, it's a matter of cognitive load in parsing out lines. If there's a big space, it's easy to see it. Yes, it's subjective, but I think ormolu is definitely putting a lot of people off by using 2 spaces because of this cognitive load issue.

**Comment by wygulmage** [comment]
I have bad eyes and like to use long identifiers in a half-width window in my vertically-oriented rotatable screen. (I'm seriously doing it right now in the window to the right of this one.)

So I use 3-space indentation, the perfect compromise that has all the advantages and none of the problems of both 2- and 4-space indentation (except convention).

**Comment by torgeirsh** [comment]
I had a crazy idea that I'm not sure is viable, but how difficult would it be to add a ligature that converts two spaces to four spaces in your favourite coding font, similar to those fonts that replace sequences like -> with a real arrow?

**Comment by cblp** [comment]
@torgeirsh This would break vertical alignment of equations. A better option is to indent with tabs.

**Comment by parsonsmatt** [comment]
honestly would find it very funny for there to be a fourk providing four space indent named `fourmolu`

**Comment by parsonsmatt** [comment]
i have released [`fourmolu`](https://hackage.haskell.org/package/fourmolu)

**Comment by afcowie** [comment]
> i have released [`fourmolu`](https://hackage.haskell.org/package/fourmolu)

Amusing, but I wish you hadn't done that (and @parsonsmatt could you at least use GitHub's fork mechanism so we can see the ahead/behind in-band?)

@mrkkrp @int-index are you absolutely sure we can't offer the four-space crowd some love here? I fully support the "zero configuration" ambition but this is the one area where we're getting significant angst that is othewise inhibiting strong adoption.

**Comment by parsonsmatt** [comment]
> (and @parsonsmatt could you at least use GitHub's fork mechanism so we can see the ahead/behind in-band?)

Forks don't have their own issue trackers, and I wanted to make the ["add a config file" issue with a "good first issue" tag](https://github.com/parsonsmatt/fourmolu/issues/2). `ormolu` maintainers are welcome to link folks dissatisfied with the goals of the project to the more permissive `fourmolu` page.

**Comment by afcowie** [comment]
> Forks don't have their own issue trackers

Actually the do! (or can). Go to a fork, then Settings, and you can tick "Issues". It's the next one underneath "Wiki". I just tried it with a couple of my forks both of personal projects and from other organizations.

**Comment by mrkkrp** [comment]
I'm going to close this one now. I do not see how the partisans of 4 space indentation are going to ever gain that critical mass to change the status quo. We released 0.1.0.0 and the style is not going to change much now. I know many companies that adopted Ormolu as their formater for Haskell. They do not complain about 2 space indentation, so for majority of people it is not an issue.

Besides, there is now `fourmolu` which, with some vigil from @parsonsmatt will continue to be a viable solution for the people who prefer 4 space indentation.
